securesystemslib/signer/_key.py

```diff
@@ -0,0 +1,199 @@
+            of the key.
+        keytype: Key type, e.g. "rsa", "ed25519" or "ecdsa-sha2-nistp256".
+        scheme: Signature scheme. For example:
+            "rsassa-pss-sha256", "ed25519", and "ecdsa-sha2-nistp256".
+        keyval: Opaque key content
+        unrecognized_fields: Dictionary of all attributes that are not managed
+            by Securesystemslib
+
+    Raises:
+        TypeError: Invalid type for an argument.
+    """
+
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(
+        self,
+        keyid: str,
+        keytype: str,
+        scheme: str,
+        keyval: Dict[str, Any],
+        unrecognized_fields: Optional[Dict[str, Any]] = None,
+    ):
+        if not all(
+            isinstance(at, str) for at in [keyid, keytype, scheme]
+        ) or not isinstance(keyval, dict):
+            raise TypeError("Unexpected Key attributes types!")
+        self.keyid = keyid
+        self.keytype = keytype
+        self.scheme = scheme
+        self.keyval = keyval
+        if unrecognized_fields is None:
+            unrecognized_fields = {}
+
+        self.unrecognized_fields = unrecognized_fields
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Key):
+            return False
+
+        return (
+            self.keyid == other.keyid
+            and self.keytype == other.keytype
+            and self.scheme == other.scheme
+            and self.keyval == other.keyval
+            and self.unrecognized_fields == other.unrecognized_fields
+        )
+
+    @classmethod
+    @abc.abstractmethod
+    def from_dict(cls, keyid: str, key_dict: Dict[str, Any]) -> "Key":
+        """Creates ``Key`` object from a serialization dict
+
+        Key implementations must override this factory constructor that is used
+        as a deserialization helper.
+
+        Users should call Key.from_dict(): it dispacthes to the actual subclass
+        implementation based on supported keys in KEY_FOR_TYPE_AND_SCHEME.
+
+        Raises:
+            KeyError, TypeError: Invalid arguments.
+        """
+        keytype = key_dict.get("keytype")
+        scheme = key_dict.get("scheme")
+        if (keytype, scheme) not in KEY_FOR_TYPE_AND_SCHEME:
+            raise ValueError(f"Unsupported public key {keytype}/{scheme}")
+
+        # NOTE: Explicitly not checking the keytype and scheme types to allow
+        # intoto to use (None,None) to lookup GPGKey, see issue #450
+        key_impl = KEY_FOR_TYPE_AND_SCHEME[(keytype, scheme)]  # type: ignore
+        return key_impl.from_dict(keyid, key_dict)
+
+    def to_dict(self) -> Dict[str, Any]:
+        """Returns a serialization dict.
+
+        Key implementations may override this method. This method is a
+        deserialization helper.
+        """
+        return {
+            "keytype": self.keytype,
+            "scheme": self.scheme,
+            "keyval": self.keyval,
+            **self.unrecognized_fields,
+        }
```

Comment:
I think it would be safer to raise NotImplementedError or even just not implement this here and require subclasses to implement their own `to_dict`. I am worried that a default `to_dict` could lead to an undetected wrong serialization of a subclass.

Reply:
sure, wfm